crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,406 @@
+            self.index.index_txout(outpoint, txout);
+        }
+
+        self.graph.apply_additions(graph_additions);
+
+        if let Some(height) = last_height {
+            self.last_height = height;
+        }
+    }
+
+    fn insert_height_internal(&mut self, tip: u32) -> Option<u32> {
+        if self.last_height < tip {
+            self.last_height = tip;
+            Some(tip)
+        } else {
+            None
+        }
+    }
+
+    /// Insert a block height that the chain source has scanned up to.
+    pub fn insert_height(&mut self, tip: u32) -> IndexedAdditions<A, I::Additions>
+    where
+        I::Additions: Default,
+    {
+        IndexedAdditions {
+            last_height: self.insert_height_internal(tip),
+            ..Default::default()
+        }
+    }
+
+    /// Insert a `txout` that exists in `outpoint` with the given `observation`.
+    pub fn insert_txout(
+        &mut self,
+        outpoint: OutPoint,
+        txout: &TxOut,
+        observation: ObservedIn<A>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let last_height = match &observation {
+            ObservedIn::Block(anchor) => self.insert_height_internal(anchor.anchor_block().height),
+            ObservedIn::Mempool(_) => None,
+        };
+
+        IndexedAdditions {
+            graph_additions: {
+                let mut graph_additions = self.graph.insert_txout(outpoint, txout.clone());
+                graph_additions.append(match observation {
+                    ObservedIn::Block(anchor) => self.graph.insert_anchor(outpoint.txid, anchor),
+                    ObservedIn::Mempool(seen_at) => {
+                        self.graph.insert_seen_at(outpoint.txid, seen_at)
+                    }
+                });
+                graph_additions
+            },
+            index_additions: <I as TxIndex>::index_txout(&mut self.index, outpoint, txout),
+            last_height,
+        }
+    }
+
+    pub fn insert_tx(
+        &mut self,
+        tx: &Transaction,
+        observation: ObservedIn<A>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let txid = tx.txid();
+
+        let last_height = match &observation {
+            ObservedIn::Block(anchor) => self.insert_height_internal(anchor.anchor_block().height),
+            ObservedIn::Mempool(_) => None,
+        };
+
+        IndexedAdditions {
+            graph_additions: {
+                let mut graph_additions = self.graph.insert_tx(tx.clone());
+                graph_additions.append(match observation {
+                    ObservedIn::Block(anchor) => self.graph.insert_anchor(txid, anchor),
+                    ObservedIn::Mempool(seen_at) => self.graph.insert_seen_at(txid, seen_at),
+                });
+                graph_additions
+            },
+            index_additions: <I as TxIndex>::index_tx(&mut self.index, tx),
+            last_height,
+        }
+    }
+
+    pub fn filter_and_insert_txs<'t, T>(
```

Comment:
```suggestion
    pub fn insert_relevant_txs<'t, T>(
```

I see this is where you need the `AddAssign` bound. Looks like you'll have to create the `Append` trait.